zhusuan/distributions/special.py

```diff
@@ -157,3 +158,77 @@ def _prob(self, given):
             return (2 * prob - 1) * inf_dtype
         else:
             return tf.cast(prob, tf.float32)
+
+
+class FlowDistribution(Distribution):
+    """
+    The class of FlowDistribution distribution.
+    The distribution describes variable which is sampled from a base
+    distribution and then is passed through an invertible function.
+    See :class:`~zhusuan.distributions.base.Distribution` for details.
+
+    :param name: A string. The name of the `StochasticTensor`. Must be unique
+        in the `BayesianNet` context.
+    :param base: An instance of `Distribution` parametrizing the base distribution.
+    :param forward: A forward function which describes how we transform the samples
+        from the base distribution. The signature of the function should be:
+            transformed, log_det = forward(base_samples)
```

Comment:
We should restrict how the forward function works. Does it apply transformation on each value (of `value_shape`) individually, or does it apply transformation on the group of values (of shape `batch_shape[-group_ndims:] + value_shape`)? There needs some careful design.

Reply:
So, I think that the function signature should be on each value independently, but to work with a batch of those as inputs. Since the flow distribution is defined as a transformation of each variable it is a deterministic function in each value independently. As long as the function can work with a batch I think its fine. Potentially we might want to flatten all the batch dimensions when it is inputted to forward and then reshaping back to make it easier for the implementators to not have the option of multiple batch dimensions.